resharper/src/CSharp/Completions/GodotSmartCompletion.cs

```diff
@@ -0,0 +1,130 @@
+using JetBrains.Annotations;
+using JetBrains.DocumentModel;
+using JetBrains.ProjectModel;
+using JetBrains.ReSharper.Feature.Services.CodeCompletion;
+using JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.LookupItems;
+using JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.LookupItems.Impl;
+using JetBrains.ReSharper.Feature.Services.CodeCompletion.Infrastructure.Match;
+using JetBrains.ReSharper.Feature.Services.CSharp.CodeCompletion.Infrastructure;
+using JetBrains.ReSharper.Feature.Services.Lookup;
+using JetBrains.ReSharper.Features.Intellisense.CodeCompletion.CSharp.Rules;
+using JetBrains.ReSharper.Plugins.Godot.ProjectModel;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.Resources;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.TextControl;
+using JetBrains.UI.Icons;
+
+namespace JetBrains.ReSharper.Plugins.Godot.CSharp.Completions
+{
+    [Language(typeof(CSharpLanguage))]
+    public class GodotSmartCompletion : CSharpItemsProviderBase<CSharpCodeCompletionContext>
+    {
+        protected override bool IsAvailable(CSharpCodeCompletionContext context)
+        {
+            return context.BasicContext.CodeCompletionType == CodeCompletionType.BasicCompletion;
+        }
+
+        protected override bool AddLookupItems(CSharpCodeCompletionContext context, IItemsCollector collector)
+        {
+            if (!IsAvailable(context))
+                return false;
+
+            var project = context.NodeInFile.GetProject();
+            if (project is null)
+                return false;
+
+            if (!project.IsGodotProject())
+                return false;
+
+            var stringLiteral = context.StringLiteral();
+            if (stringLiteral is null)
+                return false;
+
+            var invocationExpression = InvocationExpressionNavigator.GetByArgument(
+                CSharpArgumentNavigator.GetByValue(context.NodeInFile.Parent as ICSharpLiteralExpression));
+
+
+            var itemsCollected = false;
+            itemsCollected |= LookupInputActions(invocationExpression, context, collector);
+            itemsCollected |= LookupNodePaths(invocationExpression, context, collector);
+
+            return itemsCollected;
+        }
+
+        private static bool LookupNodePaths(IInvocationExpression invocationExpression, CSharpCodeCompletionContext context,
+            IItemsCollector collector)
+        {
+            if (NodePathMethods.Methods.Contains(invocationExpression.InvokedMethodName()))
```

Comment:
Why we are not checking the type here? I would expect it to be `Node` or its descendants.

Reply:
You are right. I thought I had to implement a check for every Sub-Node (like `PanelContainer` etc.). But it works just with `Node`. Added the check.